File: ios/Doorbell/Texts.swift
import Foundation

final class Texts {

    private(set) var lang = "ja"

    private var config: [String: Any]?
    private var overrides: [String: Any]?
    private var bundle: Bundle = .main

    func setConfig(_ cfg: [String: Any]?) {
        config = cfg
        reload()
    }

    func setLang(_ l: String) {
        lang = l.isEmpty ? "ja" : l
        reload()
    }

    private func reload() {
        overrides = ((config?["i18n_overrides"] as? [String: Any])?[lang]) as? [String: Any]
        if let path = Bundle.main.path(forResource: lang, ofType: "lproj"),
           let b = Bundle(path: path) {
            bundle = b
        } else {
            bundle = .main
        }
    }

    func t(_ key: String, _ args: CVarArg...) -> String {
        if let ov = overrides?[key] as? String, !ov.isEmpty {
            return Texts.fillNamed(ov, args)
        }
        let fmt = bundle.localizedString(forKey: key, value: key, table: nil)
        return args.isEmpty ? fmt : String(format: fmt, arguments: args)
    }

    private static func fillNamed(_ s: String, _ args: [CVarArg]) -> String {
        guard !args.isEmpty,
              let re = try? NSRegularExpression(pattern: "\\{[A-Za-z_][A-Za-z0-9_]*\\}")
        else { return s }
        var out = s
        var i = 0
        while i < args.count,
              let m = re.firstMatch(in: out, range: NSRange(out.startIndex..., in: out)),
              let r = Range(m.range, in: out) {
            out.replaceSubrange(r, with: "\(args[i])")
            i += 1
        }
        return out
    }

    static func langDisplayName(_ lang: String) -> String {
        switch lang {
        case "ja": return "日本語"
        case "en": return "English"
        case "zh": return "中文"
        default: return lang
        }
    }
}

File: android/app/src/main/java/jp/ox/doorbell/Texts.kt
// Resolve configuration overrides before localized Android resources. Named placeholders use
// the same encounter order as the generated positional placeholders from i18n/strings.yaml.
package jp.ox.doorbell

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import java.util.Locale
import org.json.JSONObject

class Texts(private val base: Context) {

    /** Current display language: visitor language at a door, boot.ui_lang indoors. */
    var lang: String = "ja"
        private set

    private var config: JSONObject? = null
    private var overrides: JSONObject? = null
    private var res: Resources = base.resources

    /** Replace the configuration tree at startup or after config_changed. */
    fun setConfig(cfg: JSONObject?) {
        config = cfg
        reload()
    }

    /** Switch display language and its built-in resource context. */
    fun setLang(l: String) {
        lang = if (l.isEmpty()) "ja" else l
        reload()
    }

    private fun reload() {
        overrides = config?.optJSONObject("i18n_overrides")?.optJSONObject(lang)
        res = localizedResources(lang)
    }

    private fun localizedResources(l: String): Resources = try {
        val c = Configuration(base.resources.configuration)
        c.setLocale(Locale(l))
        base.createConfigurationContext(c).resources
    } catch (_: Exception) {
        base.resources
    }

    /**
     * key identifies an i18n/strings.yaml entry for override lookup; resId selects the built-in
     * translation. Arguments fill named overrides by encounter order and positional resources.
     */
    fun t(key: String, resId: Int, vararg args: Any): String {
        val ov = overrides?.optString(key).orEmpty()
        if (ov.isNotEmpty()) return fillNamed(ov, args)
        return try {
            if (args.isEmpty()) res.getString(resId) else res.getString(resId, *args)
        } catch (_: Exception) {
            key
        }
    }

    private fun fillNamed(s: String, args: Array<out Any>): String {
        if (args.isEmpty()) return s
        var i = 0
        return PLACEHOLDER.replace(s) { m -> if (i < args.size) args[i++].toString() else m.value }
    }

    companion object {
        private val PLACEHOLDER = Regex("\\{[A-Za-z_][A-Za-z0-9_]*\\}")

        /** Native language names let visitors identify their language. */
        fun langDisplayName(lang: String): String = when (lang) {
            "ja" -> "日本語"
            "en" -> "English"
            "zh" -> "中文"
            else -> lang
        }
    }
}
